aesd-char-driver/aesd-circular-buffer.c

```diff
@@ -0,0 +1,130 @@
+/**
+ * @file aesd-circular-buffer.c
+ * @brief Functions and data related to a circular buffer imlementation
+ *
+ * @author Dan Walkes
+ * @date 2020-03-01
+ * @copyright Copyright (c) 2020
+ *
+ */
+
+#ifdef __KERNEL__
+#include <linux/string.h>
+#else
+#include <string.h>
+#endif
+
+#include <stdio.h>
+#include "aesd-circular-buffer.h"
+
+
+/*
+ * @brief Private function used only by this .c file.
+ *
+ * Given the pointer of the queue, this function calculates next pointer value.
+ *
+ * @param ptr (in) - ptr of the messsage queue
+ *
+ * @return uint32_t (out) - next ptr value
+ */
+static uint32_t nextPtr(uint32_t ptr) {
+
+  if ((ptr + 1) >= AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED) {
+    return 0;
+  }
+
+  return (ptr+1);
+} // nextPtr()
+
+
+/**
+ * @param buffer the buffer to search for corresponding offset.  Any necessary locking must be performed by caller.
+ * @param char_offset the position to search for in the buffer list, describing the zero referenced
+ *      character index if all buffer strings were concatenated end to end
+ * @param entry_offset_byte_rtn is a pointer specifying a location to store the byte of the returned aesd_buffer_entry
+ *      buffptr member corresponding to char_offset.  This value is only set when a matching char_offset is found
+ *      in aesd_buffer.
+ * @return the struct aesd_buffer_entry structure representing the position described by char_offset, or
+ * NULL if this position is not available in the buffer (not enough data is written).
+ */
+struct aesd_buffer_entry *aesd_circular_buffer_find_entry_offset_for_fpos(struct aesd_circular_buffer *buffer,
+            size_t char_offset, size_t *entry_offset_byte_rtn )
+{
+    struct aesd_buffer_entry *entry = NULL;
+    int    char_pos = char_offset;
+    int    next_out_index = 0;
+
+    if (buffer != NULL) {
+	// read and write pointers are same but full flag is not set (empty condition)
+	if ((buffer->in_offs == buffer->out_offs) && (buffer->full == 0)) {
+	    fprintf(stderr, "I shouldn't be coming here. IN:%d OUT:%d FULL:%d\n\r", buffer->in_offs, buffer->out_offs, buffer->full);
```

Comment:
I like the print idea, but use a `PDEBUG` or similar macro so this will compile for your kernel builds.